README.md
# gimp-stable-diffusion

This repository includes a GIMP plugin for communication with a stable-diffusion server and a Google colab notebook for running the server.

**More flavours of the plugin are available:**
- Free image generation without running a Colab notebook or a local server: [GIMP plugin](https://github.com/blueturtleai/gimp-stable-diffusion/tree/main/stablehorde)
- Image generation with a local server: [GIMP plugin](https://github.com/blueturtleai/gimp-stable-diffusion/tree/main/local)

Please check HISTORY.md for the latest changes. 

Click here [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/blueturtleai/gimp-stable-diffusion/blob/main/gimp-stable-diffusion.ipynb) to open the notebook, if you already read the setup instructions below.

## Overview

The server exposes a REST API, which is used by the GIMP plugin to communicate with the server. Currently the plugins offers text2img, img2img and inpainting.

https://user-images.githubusercontent.com/113246030/190710535-9fb23f88-954f-4f73-afea-1475c8690754.MOV

## Manual
It doesn't exist a separate manual. Please check the following sections for installation and image generation for a detailed explanation.

## Installation
### Download files

To download the files of this repository click on "Code" and select "Download ZIP". In the ZIP you will find the file "gimp-stable-diffusion.py". This is the code for the GIMP plugin. You don't need the other files in the ZIP.

### GIMP

The plugin is tested in GIMP 2.10 and runs most likely in all 2.* releases. Excluded is 2.99, because it's already Python 3 based.

1. Start GIMP and open the preferences dialog via edit/preferences and scroll down to "folders". Expand "folders" and click on "plug-ins". Select the folder which includes your username and copy the path. 

2. Open the file explorer, navigate to this directory and copy the file "gimp-stable-diffusion.py" from the repository into this directory. If you are on MacOS or Linux, change the file permissions to 755.

3. Restart GIMP. You should now see the new menu "AI". If you don't see this, something went wrong. Please check in this case "Troubleshooting/GIMP" for possible solutions. The menu has one item "Stable Colab". This item can't currently be selected. This only works, when you opened an image before. More about this, when the server is running.

### Stable-Diffusion server
#### Prerequisites
You need a Google account and on huggingface.co. Google is needed for running a colab server and huggingface for downloading the model file. Details follow below.

#### Model file
The model file includes all the data which is needed for stable-diffusion to generate the images.
1. Create an account on https://huggingface.co. 

2. Nagivate here https://huggingface.co/CompVis/stable-diffusion-v-1-4-original and agree to the shown agreement. 

3. Go to "Settings/Access Tokens" on the left side.

4. Click on "New Token", enter a name, select "Read" as role, click on create and copy the token.

#### Colab server
1. Open this link in a new tab [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/blueturtleai/gimp-stable-diffusion/blob/main/gimp-stable-diffusion.ipynb)

2. Click on "connect" and wait until the status changes to "connected".

3. Click on the arrow left to "NVIDIA GPU" and wait until you see a checkmark on the left.

4. Click on the arrow left to "Mount Google Drive" and confirm the mount. Wait until you see a checkmark on the left.

5. Set the model path. It is recommended to use the default path. That way you don't have to adjust the path manually every time. If the model file doesn't exist at this location, it is automatically downloaded from Huggingface. Please make sure you have enough free space on your drive (about 4 GB). If you choose an individual path, please read the hints in the notebook. Execute the step.

6. Execute the step "Setup Environment".

7. Execute the step "Python Definitions".

8. Execute the step "Select and Load Model". In the selector for the model files there is currently only one entry. When v1.5 has been released, this model will be added to the selector. If the model file doesn't exist at the location you selected above, it will automatically be downloaded from Huggingface. Please make sure you have enough free space on your drive (about 4 GB). It is necessary, that you created an account on Huggingface and accepted the terms of service. Otherwise it's not possible to download the file. You will also need the Huggingface access token you created before.

9. Click on the arrow left to "Waiting for GIMP requests". The arrow on the left won't stop spinning in this case. If everything is okay, you should see something like this:
```
 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off

INFO:werkzeug: * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)

 * Running on https://*.trycloudflare.com <- copy this URL
 * Traffic stats available on http://127.0.0.1:4040
 ```
 
10. Copy the URL from above, which reads like ```https://*.trycloudflare.com```. This is the URL, which is used for the communication between the GIMP plugin and the server. 

## Generate images
Now we are ready for generating images.

1. Start GIMP and open an image or create a new one. It is recommended, that the image size is not larger than 512x512 as the model has been trained on this size. If you want to have larger images, use an external upscaler instead. The generated image will have the dimensions of the init image. But it may be resized to make sure, that the dimensions are a multiple of 64. The larger the image, the longer it takes to generate it and the more GPU ressources and RAM is used. If it is too larger, you will run out of memory.

2. Select the new AI/Stable Colab menu item. A dialog will open, where you can enter the details for the image generation.

   - **Generation Mode:** 
      - **Text -> Image:** Generate an image based on your prompt.
      - **Image -> Image:** Generate an image based on an init image and on your prompt.
      - **Inpainting:** Erase a part of an image and generate a new image which has the erased part filled. The erased part is filled based on your prompt. Please read the section "Inpainting" below for an explanation how inpainting works.
    
   - **Init Strength:** How much the AI should take the init image into account. The higher the value, the more will the generated image look like the init image. 0.3 is a good value to use.

   - **Prompt Strength:** How much the AI should follow the prompt. The higher the value, the more the AI will generate an image which looks like your prompt. 7.5 is a good value to use.

   - **Steps:** How many steps the AI should use to generate the image. The higher the value, the more the AI will work on details. But it also means, the longer the generation takes and the more the GPU is used. 50 is a good value to use.

   - **Seed:** This parameter is optional. If it is empty, a random seed will be generated on the server. If you use a seed, the same image is generated again in the case the same parameters for init strength, steps, etc. are used. A slightly different image will be generated, if the parameters are modified. You find the seed in an additional layer at the top left. 

   - **Number of images:** Number of images, which are created in one run. The more images you create, the more server ressources will be used and the longer you have to wait until the generated images are displayed in GIMP.
   
   - **Prompt:** How the generated image should look like.

   - **Backend root URL:** Insert the trycloudflare.com URL you copied from the server. The URL should look like this ```https://*.trycloudflare.com```.

3. Click on the OK button. The values you inserted into the dialog and the init image will be transmitted to the server, which starts now generating the image. On the colab browser tab you can see what's going on. When the image has been generated successfully, it will be shown as a new image in GIMP. The used seed is shown at the top left in an additional layer.

### Inpainting
Inpainting means replacing a part of an existing image. For example if you don't like the face on an image, you can replace it. **Inpainting is currently still in experimental stage. So, please don't expect perfect results.** The experimental stage is caused by the server side and not by GIMP.

For inpainting it's necessary to prepare the input image because the AI needs to know which part you want to replace. For this purpose you replace this image part by transparency. To do so, open the init image in GIMP and select "Layer/Transparency/Add alpha channel". Select now the part of the image which should be replaced and delete it. You can also use the eraser tool. 

For the prompt you use now a description of the new image. For example the image shows currently "a little girl running over a meadow with a balloon" and you want to replace the balloon by a parachute. You just write now "a little girl running over a meadow with a parachute".

## Troubleshooting
### GIMP
#### AI menu is not shown
##### Linux
   - If you get this error ```gimp: LibGimpBase-WARNING: gimp: gimp_wire_read(): error```, it's very likely, that you have a GIMP version installed, which doesn't include Python. Check, if you have got the menu "Filters > Python-Fu > Console". If it is missing, please install GIMP from here: https://flathub.org/apps/details/org.gimp.GIMP.

##### macOS
   - Please double check, if the permissions of the plugin py file are set to 755. It seems, that changing permissions doesn't work via the file manager. Please open a terminal, cd to the plugins directory and run "chmod ugo+x *py".
   
##### macOS/Linux
   - Open a terminal an try to run the plugin py file manually via ```python <path-to-plugin-folder>/gimp-stable-diffusion.py```. You should see the error message, that "gimpfu" is unknown. Make sure, that you are running Python 2, as this version is used by GIMP. If other errors occur, please reinstall GIMP.

### Colab server
#### Ressource limits
The ressources on the colab server are limited. So, it's a good idea to stop it when you don't use it. 
   - If you only don't use it for a short time, just stop the last step (Waiting for GIMP requests). To do so, click on the spinning circle on the left. If you want to use it again, just execute the last step again. The URL for accessing the server will be different, so copy it again.

   - If you don't use if for a longer time, the best is to release all ressources. To do so, select "Runtime/Disconnect and delete runtime". If you want to use it again, you have to start again at step 1.

If you generated several images, the ressources of the colab server will be exhausted at some point. This happens pretty quickly, if you use the free plan. It takes longer for the pro plans. If this happens, an error will occur and you have to wait for some time until you can generate images again.

## FAQ

**Will GIMP 3 be supported?** 
Yes, the plugin will be ported to GIMP 3.

**Does it run locally?** According to Google it should be possible to run the notebook locally. As I don't have a local GPU, I can't try it myself. If you give it a try, I would be happy to know, if it really works.

**Will Out-Painting be supported?** I first need to check the details and see what's possible. So, unfortunately I can't promise it currently.

**Can the official stable-diffusion API be used?** Unfortunately, this is currently not possible. The reason is, that this API currently can only be accessed via gRPC and it's not possible to use this protocol in a GIMP plugin. As soon as the API is available as a REST API, it will be possible to port the plugin.

**How do I report an error or request a new feature?** Please open a new issue in this repository.




HISTORY.md
# History
## GIMP Plugin
### 1.2.0
#### Changes
- Now text2img is supported.
- The dialog has been optimized.

### 1.1.2
#### Changes
- Exception handling further improved

### 1.1.1
#### Changes
- Data is now transferred to the server using https

### 1.1.0
#### Changes
- Inpainting is now supported

### 1.0.2
#### Changes
- **Image count**: It is now possible to create up to 4 images in one run

- **Seed**: Can now be blank if not used (before -1)

- **Backend URL**: Needs no longer to end with "/"

- **Exception handling**: Has been improved

### 1.0.1
#### CHANGES
- **API**: Data is now transferred as JSON. It is checked, if the API version of the client and the server matches. If mismatch, user is requested to update plugin and/or server. 

- **Seed**: The seed of the generated image is now displayed in an additional layer. It can be passed to the server to generate the same image again. This only works, if the same parameters (init strength, number of steps, etc) are used.

### 1.0.0
Initial version

## Stable-Diffusion server
### 1.3.0
Changes
- Now text2img is supported.

### 1.2.1
#### Changes
- Stable-Diffusion Model v1.5 is now supported.

### 1.2.0
#### Changes
- It's no longer necessary to upload model files yourself to your Google drive. If a model file is missing, it is automatically downloaded from Huggingface. A Huggingface account and acceptance of the terms of service is still needed.

### 1.1.2
#### Changes
- Instead of ngrok now cloudflare is used as the tunnel provider. The switch is done because several users reported problems. Additional benefit is, that it's no longer necessary to register on ngrok and enter the authkey. Thanks for suggesting this @opencoca.

### 1.1.1
#### Changes
- Mounting Google drive and setting model path are now two separated steps

### 1.1.0
#### Changes
- The notebook is now based on the deforum notebook (before pharmapsychotic)
- Inpainting is now supported

### 1.0.3
#### Changes
- **Image count**: It is now possible to create up to 4 images in one run

### 1.0.2
#### Changes
- **API**: Data is now transferred in JSON. It is checked, if API version of the client and server matches. If mismatch, code 405 is returned.

- **Seed**: If a seed != -1 is transferred to the server, this seed is used for image generation.

### 1.0.1
#### Changes
- **Image dimensions**: The image dimensions are now no longer fixed to 512x512. The dimensions of the init image are now used instead. The dimensions are adjusted to a multiple of 64.

### 1.0.0
Initial version


gimp-stable-diffusion.py
#!/usr/bin/python

# v1.1.2

import urllib2
import tempfile
import os
import base64
import json
import re
import ssl

from gimpfu import *

INIT_FILE = "init.png"
GENERATED_FILE = "generated.png"
API_ENDPOINT = "api/img2img"
API_VERSION = 4

HEADER_ACCEPT = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
HEADER_USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0"

headers = {"Accept": HEADER_ACCEPT, "User-Agent": HEADER_USER_AGENT, "Content-Type": "application/json"}
ssl._create_default_https_context = ssl._create_unverified_context

initFile = r"{}".format(os.path.join(tempfile.gettempdir(), INIT_FILE))
generatedFile = r"{}".format(os.path.join(tempfile.gettempdir(), GENERATED_FILE))

def getImageData(image, drawable):
   pdb.file_png_save_defaults(image, drawable, initFile, initFile)
   initImage = open(initFile, "rb")
   encoded = base64.b64encode(initImage.read())
   return encoded

def displayGenerated(images):
   color = pdb.gimp_context_get_foreground()
   pdb.gimp_context_set_foreground((0, 0, 0))

   for image in images:
      imageFile = open(generatedFile, "wb+")
      imageFile.write(base64.b64decode(image["image"]))
      imageFile.close()

      imageLoaded = pdb.file_png_load(generatedFile, generatedFile)
      pdb.gimp_display_new(imageLoaded)
      # image, drawable, x, y, text, border, antialias, size, size_type, fontname
      pdb.gimp_text_fontname(imageLoaded, None, 2, 2, str(image["seed"]), -1, TRUE, 12, 1, "Sans")
      pdb.gimp_image_set_active_layer(imageLoaded, imageLoaded.layers[1])

   pdb.gimp_context_set_foreground(color)
   return

def img2img(image, drawable, isInpainting, maskBrightness, maskContrast, initStrength, promptStrength, steps, seed, imageCount, prompt, url):
   data = {
      "inpainting": bool(isInpainting),
      "mask_brightness": float(maskBrightness),
      "mask_contrast": float(maskContrast),
      "init_strength": float(initStrength),
      "prompt_strength": float(promptStrength),
      "steps": int(steps),
      "width": int(image.width),
      "height": int(image.height),
      "prompt": prompt,
      "image_count": int(imageCount),
      "api_version": API_VERSION
   }

   seed = -1 if not seed else int(seed)
   data.update({"seed": seed})

   imageData = getImageData(image, drawable)
   data.update({"init_img": imageData})
   data = json.dumps(data)

   url = url + "/" if not re.match(".*/$", url) else url
   url = re.sub("http://", "https://", url)
   url = url + API_ENDPOINT

   request = urllib2.Request(url=url, data=data, headers=headers)
   pdb.gimp_progress_set_text("starting dreaming now...")

   try:
      response = urllib2.urlopen(request)
      data = response.read()

      try:
         data = json.loads(data)
         displayGenerated(data["images"])

         if os.path.exists(initFile):
            os.remove(initFile)

         if os.path.exists(generatedFile):
            os.remove(generatedFile)
      except Exception as ex:
         raise Exception(data)

   except Exception as ex:
      if isinstance(ex, urllib2.HTTPError) and ex.code == 405:
         raise Exception("GIMP plugin and stable-diffusion server don't match. Please update the GIMP plugin. If the error still occurs, please reopen the colab notebook.")
      else:
         raise ex

   return

register(
   "img2img",
   "img2img",
   "img2img",
   "BlueTurtleAI",
   "BlueTurtleAI",
   "2022",
   "<Image>/AI/Stable img2img",
   "*",
   [
      (PF_TOGGLE, "isInpainting", "Inpainting", False),
      (PF_SLIDER, "maskBrightness", "Inpainting\nMask Brightness", 1.0, (0.0, 1.0, 0.1)),
      (PF_SLIDER, "maskContrast", "Inpainting\nMask Contrast", 1.0, (0.0, 1.0, 0.1)),
      (PF_SLIDER, "initStrength", "Init Strength", 0.3, (0.0, 1.0, 0.1)),
      (PF_SLIDER, "promptStrength", "Prompt Strength", 7.5, (0, 20, 0.5)),
      (PF_SLIDER, "steps", "Steps", 50, (10, 150, 1)),
      (PF_STRING, "seed", "Seed (optional)", ""),
      (PF_SLIDER, "imageCount", "Number of images", 1, (1, 4,1)),
      (PF_STRING, "prompt", "Prompt", ""),
      (PF_STRING, "url", "Backend root URL", "")
   ],
   [],
   img2img
)

main()


local/gimp-stable-diffusion-local.py
#!/usr/bin/python

# v1.1.0

import urllib2
import tempfile
import os
import base64
import json
import re
import random
import math
import gimp
from gimpfu import *

VERSION = 110
INIT_FILE = "init.png"
GENERATED_FILE = "generated.png"
API_ENDPOINT = "predictions"

initFile = r"{}".format(os.path.join(tempfile.gettempdir(), INIT_FILE))
generatedFile = r"{}".format(os.path.join(tempfile.gettempdir(), GENERATED_FILE))

def checkUpdate():
  try:
     gimp.get_data("update_checked")
     updateChecked = True
  except Exception as ex:
     updateChecked = False

  if updateChecked is False:
     try:
        url = "https://raw.githubusercontent.com/blueturtleai/gimp-stable-diffusion/main/local/version.json"
        response = urllib2.urlopen(url)
        data = response.read()
        data = json.loads(data)
        gimp.set_data("update_checked", "1")

        if VERSION < int(data["version"]):
           pdb.gimp_message(data["message"])
     except Exception as ex:
        ex = ex

def cleanup():
   try:
      if os.path.exists(initFile):
         os.remove(initFile)

      if os.path.exists(generatedFile):
         os.remove(generatedFile)
   except Exception as ex:
      ex = ex

def getImages(data, seed):
   images = []

   for counter in range(len(data["output"])):
      image = re.match("data:image/png;base64,(.*)", data["output"][counter]).group(1)
      image = {"img": image, "seed": seed}
      images.append(image)

   return images

def getImageData(image, drawable):
   pdb.file_png_save_defaults(image, drawable, initFile, initFile)
   initImage = open(initFile, "rb")
   encoded = base64.b64encode(initImage.read())
   return encoded

def displayGenerated(images):
   color = pdb.gimp_context_get_foreground()
   pdb.gimp_context_set_foreground((0, 0, 0))

   for image in images:
      imageFile = open(generatedFile, "wb+")
      imageFile.write(base64.b64decode(image["img"]))
      imageFile.close()

      imageLoaded = pdb.file_png_load(generatedFile, generatedFile)
      pdb.gimp_display_new(imageLoaded)
      # image, drawable, x, y, text, border, antialias, size, size_type, fontname
      pdb.gimp_text_fontname(imageLoaded, None, 2, 2, str(image["seed"]), -1, TRUE, 12, 1, "Sans")
      pdb.gimp_image_set_active_layer(imageLoaded, imageLoaded.layers[1])

   pdb.gimp_context_set_foreground(color)
   return

def generate(image, drawable, mode, initStrength, promptStrength, steps, seed, prompt, url):
   if image.width < 384 or image.width > 1024 or image.height < 384 or image.height > 1024:
      raise Exception("Invalid image size. Image needs to be between 384x384 and 1024x1024.")

   if image.width * image.height > 786432:
      raise Exception("Invalid image size. Maximum size is 1024x768 or 768x1024.")

   if prompt == "":
      raise Exception("Please enter a prompt.")

   if mode == "MODE_INPAINTING" and drawable.has_alpha == 0:
      raise Exception("Invalid image. For inpainting an alpha channel is needed.")

   pdb.gimp_progress_init("", None)

   input = {
      "prompt": prompt,
      "num_inference_steps": int(steps),
      "guidance_scale": float(promptStrength)
   }

   if image.width % 64 != 0:
      width = math.floor(image.width/64) * 64
   else:
      width = image.width

   if image.height % 64 != 0:
      height = math.floor(image.height/64) * 64
   else:
      height = image.height

   input.update({"width": int(width)})
   input.update({"height": int(height)})

   if not seed:
      seed = random.randint(0, 2**31)
   else:
      seed = int(seed)

   input.update({"seed": seed})

   if mode == "MODE_IMG2IMG" or mode == "MODE_INPAINTING":
      imageData = getImageData(image, drawable)
      input.update({"init_image": imageData})
      input.update({"prompt_strength": (1 - float(initStrength))})

   if mode == "MODE_INPAINTING":
      input.update({"mask": imageData})

   data = {"input": input}
   data = json.dumps(data)

   headers = {"Content-Type": "application/json", "Accept": "application/json"}

   url = url + "/" if not re.match(".*/$", url) else url
   url = url + API_ENDPOINT

   request = urllib2.Request(url=url, data=data, headers=headers)
   pdb.gimp_progress_set_text("starting dreaming now...")

   try:
      response = urllib2.urlopen(request)
      data = response.read()

      try:
         data = json.loads(data)
      except Exception as ex:
         raise Exception(data)

      if data["status"] == "failed":
         raise Exception("The image couldn't be generated: " + data["error"])

      images = getImages(data, seed)
      displayGenerated(images)
   except Exception as ex:
      raise ex
   finally:
      pdb.gimp_progress_end()
      cleanup()
      checkUpdate()

   return

register(
   "stable-local",
   "stable-local",
   "stable-local",
   "BlueTurtleAI",
   "BlueTurtleAI",
   "2022",
   "<Image>/AI/Stable Local",
   "*",
   [
      (PF_RADIO, "mode", "Generation Mode", "MODE_TEXT2IMG", (
         ("Text -> Image", "MODE_TEXT2IMG"),
         ("Image -> Image", "MODE_IMG2IMG"),
         ("Inpainting", "MODE_INPAINTING")
      )),
      (PF_SLIDER, "initStrength", "Init Strength", 0.3, (0.0, 1.0, 0.1)),
      (PF_SLIDER, "promptStrength", "Prompt Strength", 7.5, (0, 20, 0.5)),
      (PF_SLIDER, "steps", "Steps", 50, (10, 150, 1)),
      (PF_STRING, "seed", "Seed (optional)", ""),
      (PF_STRING, "prompt", "Prompt", ""),
      (PF_STRING, "url", "Backend root URL", "")
   ],
   [],
   generate
)

main()


stablehorde/gimp-stable-diffusion-horde.py
#!/usr/bin/python

# v1.3.2

import urllib2
import tempfile
import os
import base64
import json
import ssl
import sched, time
import math
import gimp

from gimpfu import *

VERSION = 132
INIT_FILE = "init.png"
GENERATED_FILE = "generated.png"
API_ROOT = "https://stablehorde.net/api/v2/"

# check every 5 seconds
CHECK_WAIT = 5
checkMax = None

ssl._create_default_https_context = ssl._create_unverified_context

initFile = r"{}".format(os.path.join(tempfile.gettempdir(), INIT_FILE))
generatedFile = r"{}".format(os.path.join(tempfile.gettempdir(), GENERATED_FILE))
s = sched.scheduler(time.time, time.sleep)

checkCounter = 0
id = None

def checkUpdate():
  try:
     gimp.get_data("update_checked")
     updateChecked = True
  except Exception as ex:
     updateChecked = False

  if updateChecked is False:
     try:
        url = "https://raw.githubusercontent.com/blueturtleai/gimp-stable-diffusion/main/stablehorde/version.json"
        response = urllib2.urlopen(url)
        data = response.read()
        data = json.loads(data)
        gimp.set_data("update_checked", "1")

        if VERSION < int(data["version"]):
           pdb.gimp_message(data["message"])
     except Exception as ex:
        ex = ex

def getImageData(image, drawable):
   pdb.file_png_save_defaults(image, drawable, initFile, initFile)
   initImage = open(initFile, "rb")
   encoded = base64.b64encode(initImage.read())
   return encoded

def displayGenerated(images):
   color = pdb.gimp_context_get_foreground()
   pdb.gimp_context_set_foreground((0, 0, 0))

   for image in images:
      imageFile = open(generatedFile, "wb+")
      imageFile.write(base64.b64decode(image["img"]))
      imageFile.close()

      imageLoaded = pdb.file_webp_load(generatedFile, generatedFile)
      pdb.gimp_display_new(imageLoaded)
      # image, drawable, x, y, text, border, antialias, size, size_type, fontname
      pdb.gimp_text_fontname(imageLoaded, None, 2, 2, str(image["seed"]), -1, TRUE, 12, 1, "Sans")
      pdb.gimp_image_set_active_layer(imageLoaded, imageLoaded.layers[1])

   pdb.gimp_context_set_foreground(color)
   return

def getImages():
   url = API_ROOT + "generate/status/" + id
   response = urllib2.urlopen(url)
   data = response.read()
   data = json.loads(data)

   return data["generations"]

def checkStatus():
   url = API_ROOT + "generate/check/" + id
   response = urllib2.urlopen(url)
   data = response.read()
   data = json.loads(data)

   global checkCounter
   checkCounter = checkCounter + 1

   if data["processing"] == 0:
      text = "Queue position: " + str(data["queue_position"]) + ", Wait time: " + str(data["wait_time"]) + "s"
   elif data["processing"] > 0:
      text = "Generating..."

   pdb.gimp_progress_set_text(text)

   if checkCounter < checkMax and data["done"] is False:
      if data["is_possible"] is True:
         s.enter(CHECK_WAIT, 1, checkStatus, ())
         s.run()
      else:
         raise Exception("Currently no worker available to generate your image. Please try again later.")
   elif checkCounter == checkMax:
      minutes = (checkMax * CHECK_WAIT)/60
      raise Exception("Image generation timed out after " + str(minutes) + " minutes. Please try it again later.")
   elif data["done"] == True:
      return

def generate(image, drawable, mode, initStrength, promptStrength, steps, seed, nsfw, prompt, apikey, maxWaitMin):
   if image.width < 384 or image.width > 1024 or image.height < 384 or image.height > 1024:
      raise Exception("Invalid image size. Image needs to be between 384x384 and 1024x1024.")

   if prompt == "":
      raise Exception("Please enter a prompt.")

   if mode == "MODE_INPAINTING" and drawable.has_alpha == 0:
      raise Exception("Invalid image. For inpainting an alpha channel is needed.")

   pdb.gimp_progress_init("", None)

   global checkMax
   checkMax = (maxWaitMin * 60)/CHECK_WAIT

   try:
      params = {
         "cfg_scale": float(promptStrength),
         "steps": int(steps),
         "seed": seed
      }

      data = {
         "params": params,
         "prompt": prompt,
         "nsfw": nsfw,
         "censor_nsfw": False
      }

      if image.width % 64 != 0:
         width = math.floor(image.width/64) * 64
      else:
         width = image.width

      if image.height % 64 != 0:
         height = math.floor(image.height/64) * 64
      else:
         height = image.height

      params.update({"width": int(width)})
      params.update({"height": int(height)})

      if mode == "MODE_IMG2IMG":
         init = getImageData(image, drawable)
         data.update({"source_image": init})
         data.update({"source_processing": "img2img"})
         params.update({"denoising_strength": (1 - float(initStrength))})
      elif mode == "MODE_INPAINTING":
         init = getImageData(image, drawable)
         models = ["stable_diffusion_inpainting"]
         data.update({"source_image": init})
         data.update({"source_processing": "inpainting"})
         data.update({"models": models})

      data = json.dumps(data)

      apikey = "0000000000" if not apikey else apikey

      headers = {"Content-Type": "application/json", "Accept": "application/json", "apikey": apikey}
      url = API_ROOT + "generate/async"

      request = urllib2.Request(url=url, data=data, headers=headers)

      response = urllib2.urlopen(request)
      data = response.read()

      try:
         data = json.loads(data)
         global id
         id = data["id"]
      except Exception as ex:
         raise Exception(data)

      checkStatus()
      images = getImages()
      displayGenerated(images)

   except urllib2.HTTPError as ex:
      try:
         data = ex.read()
         data = json.loads(data)

         if "message" in data:
            message = data["message"]
         else:
            message = str(ex)
      except Exception:
         message = str(ex)

      raise Exception(message)
   except Exception as ex:
      raise ex
   finally:
      pdb.gimp_progress_end()
      checkUpdate()

   return

register(
   "stable-horde",
   "stable-horde",
   "stable-horde",
   "BlueTurtleAI",
   "BlueTurtleAI",
   "2022",
   "<Image>/AI/Stablehorde",
   "*",
   [
      (PF_RADIO, "mode", "Generation Mode", "MODE_TEXT2IMG", (
         ("Text -> Image", "MODE_TEXT2IMG"),
         ("Image -> Image", "MODE_IMG2IMG"),
         ("Inpainting", "MODE_INPAINTING")
      )),
      (PF_SLIDER, "initStrength", "Init Strength", 0.3, (0, 1, 0.1)),
      (PF_SLIDER, "promptStrength", "Prompt Strength", 7.5, (0, 20, 0.5)),
      (PF_SLIDER, "steps", "Steps", 50, (10, 150, 1)),
      (PF_STRING, "seed", "Seed (optional)", ""),
      (PF_TOGGLE, "nsfw", "NSFW", False),
      (PF_STRING, "prompt", "Prompt", ""),
      (PF_STRING, "apiKey", "API key (optional)", ""),
      (PF_SLIDER, "maxWaitMin", "Max Wait (minutes)", 5, (1, 5, 1))
   ],
   [],
   generate
)

main()
